### PR TITLE
Build cleanup with go 1.10

### PR DESCRIPTION
**Pull request by thockin**
Simplify our build/test scripts now that go 1.10 is in play.  It has caching that seems to actually work.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: third_party/BUILD
Comment (ixdy) [inline third_party/BUILD]: this seems wrong, but I don't know how it got there

Comment (thockin) [reply third_party/BUILD]: update bazel did it.

Comment (ixdy) [reply third_party/BUILD]: To answer my own question: etcd v2.2 included a proto file in the `Documentation/rfc` dir, but it appears to have been removed by etcd v3.

We aren't excluding `third_party/etcd*` from gazelle/kazel, but we probably should be. (We do git ignore these directories.)

**Comment by thockin** [comment]
I deleted the BUILD change that @ixdy disagreed with, and it passes CI now.  Weird that update-bazel.sh locally still suggests that change.

EDIT: I just nuked third_party and re-checked-out, and it is fine now.  Something that was .gitignore'd I guess.

Anyway, all green now.  Does the new go cache help CI build times if we volume mount it?

**Comment by ixdy** [comment]
right, I figured out what was creating that weird update in https://github.com/kubernetes/kubernetes/pull/62151#discussion_r179336831.

cc @BenTheElder

**Comment by ixdy** [comment]
also cc @sigma

**Comment by BenTheElder** [comment]
> Anyway, all green now. Does the new go cache help CI build times if we volume mount it?

What storage would we mount? AFAICT to do this effectively you need something more sophisticated than just mounting some persistent storage to the cache dir because these caches are designed for one build of one source tree on a dev machine vs many PRs over many branches simultaneously in CI.

Previously for the bazel caching we did just mount through some storage local to each node for `$HOME/.cache/bazel` and limit jobs using it to one-per-node but this doesn't really behave well for the same reasons.

We [currently use](https://github.com/kubernetes/test-infra/tree/master/greenhouse) bazel [remote caching](https://docs.bazel.build/versions/master/remote-caching.html) support for caching which does speed up the build (significantly!) and solves these problems by being content addressed and over the network instead of on-disk.

We probably *could* build something similar to [the travis-ci cache](https://docs.travis-ci.com/user/caching/) but this doesn't exist for us currently.

**Comment by sigma** [comment]
https://github.com/golang/go/blob/master/src/cmd/go/internal/cache/cache.go#L38..L48 gives more details:

    // It is safe for multiple processes on a single machine to use the
    // same cache directory in a local file system simultaneously.
    // They will coordinate using operating system file locks and may
    // duplicate effort but will not corrupt the cache.
    //
    // However, it is NOT safe for multiple processes on different machines
    // to share a cache directory (for example, if the directory were stored
    // in a network file system). File locking is notoriously unreliable in
    // network file systems and may not suffice to protect the cache.

So it's more like multiple builds / multiple trees / one machine.
I'm not entirely sure where containers fall as far as file locking is concerned, but it seems worth exploring.
Also note that, for some reason I don't really understand, the absolute path of each source file seems to important to get cache hits, so no gain should be expected unless the path to the repo is the same in all containers.

**Comment by thockin** [comment]
Roger.  PTAL when you get a sec.  I have wanted to delete this code or years :)

**Review by cblecker — APPROVED** [review]
Less code is best code. Not LGTMing because this could use more than one set of eyes, but it looks fine to me from what I can tell.

File: build/root/Makefile.generated_files
Comment (ixdy) [inline build/root/Makefile.generated_files]: I don't understand how `make` decides when to run this rule, but I seem to be able to confuse it.

I checked out your PR locally, then ran a few commands:
```console
$ git status
On branch pr/62151
nothing to commit, working tree clean
$ make clean
+++ [0405 15:31:25] Verifying Prerequisites....          
+++ [0405 15:31:26] Removing _output directory
$ make generated_files
+++ [0405 15:31:51] Building go targets for linux/amd64: 
    ./vendor/k8s.io/code-generator/cmd/deepcopy-gen                                                                                                                                                                 
+++ [0405 15:31:57] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/defaulter-gen
+++ [0405 15:32:01] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/conversion-gen
+++ [0405 15:32:07] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/openapi-gen
+++ [0405 15:32:12] Building go targets for linux/amd64:
    ./vendor/github.com/jteeuwen/go-bindata/go-bindata
$ git status
On branch pr/62151                                       
nothing to commit, working tree clean
$ vim translations/kubectl/en_US/LC_MESSAGES/k8s.po
# edit some lines
$ git status
On branch pr/62151
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   translations/kubectl/en_US/LC_MESSAGES/k8s.po

no changes added to commit (use "git add" and/or "git commit -a")
$ git diff
diff --git a/translations/kubectl/en_US/LC_MESSAGES/k8s.po b/translations/kubectl/en_US/LC_MESSAGES/k8s.po
index 4c2f11466b..5d9a2cdf00 100644
--- a/translations/kubectl/en_US/LC_MESSAGES/k8s.po
+++ b/translations/kubectl/en_US/LC_MESSAGES/k8s.po
@@ -22,9 +22,9 @@ msgstr ""
 #: pkg/kubectl/cmd/create_clusterrolebinding.go:35
 msgid ""
 "\n"
-"\t\t  # Create a ClusterRoleBinding for user1, user2, and group1 using the "
+"\t\t  # Creat a ClusterRoleBinding for user1, user2, and group1 using the "
 "cluster-admin ClusterRole\n"
-"\t\t  kubectl create clusterrolebinding cluster-admin --clusterrole=cluster-"
+"\t\t  kubectl creat clusterrolebinding cluster-admin --clusterrole=cluster-"
 "admin --user=user1 --user=user2 --group=group1"
 msgstr ""
 "\n"
$ make generated_files
$ git diff
diff --git a/pkg/generated/bindata.go b/pkg/generated/bindata.go
index 7cbfc6154f..a657a24ae7 100644
--- a/pkg/generated/bindata.go
+++ b/pkg/generated/bindata.go
@@ -6975,9 +6975,9 @@ msgstr ""
 #: pkg/kubectl/cmd/create_clusterrolebinding.go:35
 msgid ""
 "\n"
-"\t\t  # Create a ClusterRoleBinding for user1, user2, and group1 using the "
+"\t\t  # Creat a ClusterRoleBinding for user1, user2, and group1 using the "
 "cluster-admin ClusterRole\n"
-"\t\t  kubectl create clusterrolebinding cluster-admin --clusterrole=cluster-"
+"\t\t  kubectl creat clusterrolebinding cluster-admin --clusterrole=cluster-"
 "admin --user=user1 --user=user2 --group=group1"
 msgstr ""
 "\n"
diff --git a/translations/kubectl/en_US/LC_MESSAGES/k8s.po b/translations/kubectl/en_US/LC_MESSAGES/k8s.po
index 4c2f11466b..5d9a2cdf00 100644
--- a/translations/kubectl/en_US/LC_MESSAGES/k8s.po      
+++ b/translations/kubectl/en_US/LC_MESSAGES/k8s.po      
@@ -22,9 +22,9 @@ msgstr ""                              
 #: pkg/kubectl/cmd/create_clusterrolebinding.go:35      
 msgid ""                                                
 "\n"                                                    
-"\t\t  # Create a ClusterRoleBinding for user1, user2, and group1 using the "
+"\t\t  # Creat a ClusterRoleBinding for user1, user2, and group1 using the "
 "cluster-admin ClusterRole\n"                           
-"\t\t  kubectl create clusterrolebinding cluster-admin --clusterrole=cluster-"
+"\t\t  kubectl creat clusterrolebinding cluster-admin --clusterrole=cluster-"
 "admin --user=user1 --user=user2 --group=group1"        
 msgstr ""                                               
 "\n"
$ git checkout HEAD translations/kubectl/en_US/LC_MESSAGES/k8s.po
$ make generated_files
$ git diff
diff --git a/pkg/generated/bindata.go b/pkg/generated/bindata.go
index 7cbfc6154f..a657a24ae7 100644                      
--- a/pkg/generated/bindata.go                           
+++ b/pkg/generated/bindata.go                           
@@ -6975,9 +6975,9 @@ msgstr ""                          
 #: pkg/kubectl/cmd/create_clusterrolebinding.go:35      
 msgid ""                                                
 "\n"                                                    
-"\t\t  # Create a ClusterRoleBinding for user1, user2, and group1 using the "
+"\t\t  # Creat a ClusterRoleBinding for user1, user2, and group1 using the "
 "cluster-admin ClusterRole\n"                           
-"\t\t  kubectl create clusterrolebinding cluster-admin --clusterrole=cluster-"
+"\t\t  kubectl creat clusterrolebinding cluster-admin --clusterrole=cluster-"
 "admin --user=user1 --user=user2 --group=group1"        
 msgstr ""                                               
 "\n"
```

So basically, if I update one of the input files, then make the generated files, and then revert my changes to the input file, it doesn't regenerate the output again.

File: build/root/Makefile.generated_files
Comment (thockin) [reply build/root/Makefile.generated_files]: fixed in next push

File: build/root/Makefile.generated_files
Comment (thockin) [reply build/root/Makefile.generated_files]: ah, yeah.  I tried taking a shortcut, but not too short of a shortcut.  Instead I will take the real shortcut.  fixed in next push.

**Comment by thockin** [comment]
new push is up

**Comment by BenTheElder** [comment]
thanks @cblecker. above failures are from https://github.com/kubernetes/kubernetes/issues/62356

File: build/root/Makefile.generated_files
Comment (rmmh) [inline build/root/Makefile.generated_files]: this is insufficient-- the translations bindata depends on `translations/kubectl/template.pot`, which depends on each translation messages file

Comment (rmmh) [reply build/root/Makefile.generated_files]: looking into it more: the .pot and .mo deps are never actually used, so this is fine

File: build/root/Makefile.generated_files
Comment (thockin) [reply build/root/Makefile.generated_files]: yes

**Comment by thockin** [comment]
new push is up, needs an LGTM

**Comment by cblecker** [comment]
/lgtm
/hold

Looks good to me, and works on my darwin system. Holding for @ixdy's comments.

**Comment by cblecker** [comment]
Thanks @ixdy!
/hold cancel

**Comment by ixdy** [comment]
we may want to update the `minimum_go_version` in `hack/lib/golang.sh` to go1.10, though I guess go1.9 will still produce correct code, if inefficient.

**Comment by cblecker** [comment]
@ixdy Good call. I'll open a quick follow up PR.

**Pull request merged**
